BarFiller

Morath86 opened the pull request:
Addons\BarFiller: initial commit

Morath86 committed:
Addons\BarFiller: initial commit of BarFiller v0.1.5

z16 commented on addons/barfiller/data/settings.xml:
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.1" ?>
```
Do not provide a settings file, it will be automatically generated. The data folder should generally stay empty.

z16 commented on addons/barfiller/statics.lua:
```diff
@@ -0,0 +1,195 @@
+default_settings = {
+	bar_settings = {
+		background = { -- image: Must be the EXACT path to your Windower\Addon folder
+			image = 'C:\\Program Files (x86)\\Windower4\\addons\\barfiller\\data\\bar_bg.png',
```
Do not use hard-coded paths. You can obtain the user's Windower directory with the `windower.windower_path` variable, and the current addon's directory with `windower.addon_path`. Use those instead of hard-coding the absolute path

On a side note, if you do ever need to use user-configurable options, put those in the settings, ideally users should never be asked to edit code.

Morath86 replied:
Done.

z16 commented on addons/barfiller/barfiller.lua:
```diff
@@ -0,0 +1,118 @@
+-- Addon commands
+-- Thanks to Byrth & SnickySnacks' BattleMod addon
+windower.register_event('addon command',function(...)
+	local commands = {...}
```
Use 4 spaces for indentation.

Morath86 replied:
Done.

z16 commented on addons/barfiller/barfiller.lua:
```diff
@@ -0,0 +1,118 @@
+			help()
+		end
+	else
+		print('Missing or invalid command, use "bf help" or "bf h" for list of available commands.')
```
May as well call the help function here instead of telling people to call the help function :)

Morath86 replied:
Done.

z16 commented on addons/barfiller/statics.lua:
```diff
@@ -0,0 +1,195 @@
+
+-- Default settings file
+default_settings = {
+	bar_settings = {
```
The naming convention for settings files is _PascalCase_, i.e. capitalize the first letter of each word in a multi-word identifier, without underscores.

Morath86 replied:
Done.

z16 commented on addons/barfiller/statics.lua:
```diff
@@ -0,0 +1,195 @@
+}
+
+-- Current Game Resolution
+windowerResX = windower.get_windower_settings().x_res
```
The naming convention in code is to use _snake_case_, i.e. all lower cased letters with underscores separating single words in an identifier.

Morath86 replied:
Done.

z16 commented on addons/barfiller/barfiller.lua:
```diff
@@ -0,0 +1,118 @@
+        xp.tnl = xp.total - xp.current
+		calcExpBarPerc()
+	end
+end)
```
The world would be a better place if every file ended with a new line :)

Morath86 replied:
Done.

z16 commented on addons/barfiller/barfiller.lua:
```diff
@@ -0,0 +1,118 @@
+windower.register_event('incoming chunk',function(id,org,modi,is_injected,is_blocked)
+	if is_injected then return end
+	if id == 0x2D then
+        local val = org:unpack('I',0x11)
```
You may consider using the `packets` library here. The advantage is that if packets ever change, the library will be the first to be adjusted, and all addons using it will automatically work again, without having to be touched.

(And it's also prettier to look at than manually extracted bits and bytes.)

Morath86 replied:
Will investigate in future update.

smd111 replied:
this is what you should need 

```
windower.register_event('incoming chunk',function(id,org,modi,is_injected,is_blocked)
    if is_injected then return end
    if id == 0x2D then
        local packet = packets.parse('incoming', data)
        exp_msg(packet['Param 1'],packet['Message'])
    elseif id == 0x61 then
        local packet = packets.parse('incoming', data)
        xp.current = packet['Current EXP']
        xp.total = packet['Required EXP']
        xp.tnl = xp.total - xp.current
        calcExpBarPerc()
    end
end) 
```

just so you know i have not tested this yet

Morath86 replied:
That actually worked, thank you for that!

Morath86 committed:
settings.xml is now auto-generated, replaced hard-coded paths with windower.windower_path, converted tabs to spaces, removed command err message, corrected naming conventions, files end with new line

Morath86 committed:
Thanks to smd111: Now parses packets for Current XP and Total XP

Morath86 commented:
Changed console output to chat window output

z16 commented on addons/barfiller/barfiller.lua:
```diff
@@ -0,0 +1,88 @@
+
+-- Generate Settings Files
+-- Thanks to Byrth & SnickySnacks' BattleMod addon
+settings_file = file.new('data\\settings.xml',true)
```
You never use this variable. If you just want to create a file, without using the variable, you don't have to assign it, you can just do:

``` lua
file.new('<filename>', true)
```

That said, in this case you don't need it at all. The config library will automatically create both the data folder and the settings file, so you can safely remove this line.

Morath86 replied:
Thank you for the clear explanations, I have removed that line at this time.

z16 commented on addons/barfiller/barfiller.lua:
```diff
@@ -0,0 +1,88 @@
+-- Generate Settings Files
+-- Thanks to Byrth & SnickySnacks' BattleMod addon
+settings_file = file.new('data\\settings.xml',true)
+settings_table = config.load('data\\settings.xml',DefaultSettings)
```
Since `data/settings.xml` is the default location you don't need to specify it. In fact, you probably _shouldn't_ specify it, unless you want it to be something other than the default. That way the config library is free to make its own assumptions about what and where it should be and is one less thing the addon developer needs to consider.

z16 commented on addons/barfiller/statics.lua:
```diff
@@ -0,0 +1,232 @@
+]]
+
+-- Default settings file
+DefaultSettings = {
```
Earlier when I said the settings should be _PascalCase_ I didn't mean the settings table itself, but the settings values. I.e.:

``` lua
defaults = {
    Bars = {
        Background = {
            Image = ...,
            Pos = {
                X = 164,
                Y = 6,
            },
        },
    },
}
```

Morath86 replied:
Figured I misinterpreted that. Thanks again.

z16 commented on addons/barfiller/statics.lua:
```diff
@@ -0,0 +1,232 @@
+DefaultSettings = {
+    bars = {
+        background = {
+            image = windower.windower_path..'addons\\barfiller\\data\\bar_bg.png',
```
While this works now I'm still not sure if it makes much sense to put this into the settings. Do you want users to be able to adjust it? I don't think many of them will actually change the images themselves.

If that actually is what you intended, I think there's a better way by just making a `Windower/addons/BarFiller/data/images/` folder which users can use to put their own files in to override the original files which should reside outside of the data folder. In general the addon should be completely operational when users delete their data folder, and currently it wouldn't be, because it would delete the images.

z16 commented on addons/barfiller/statics.lua:
```diff
@@ -0,0 +1,232 @@
+    bars = {
+        background = {
+            image = windower.windower_path..'addons\\barfiller\\data\\bar_bg.png',
+            pos = {
```
Is this necessary? I took from your forum post that it automatically centers the addon, so using a positional value doesn't seem to make much sense. Unless you add another setting `Center` and default it to `true`. When it's `true`, it can automatically center it, otherwise it can use what the user specifies under `Pos`.

Morath86 replied:
I thought to myself it was a good idea to put the smallest values for 800x600 resolution just as a fail safe in case the math somehow broke.

When I get around to allowing moving the users to change the `Pos` in the commands I'll fix this up.

z16 commented on addons/barfiller/statics.lua:
```diff
@@ -0,0 +1,232 @@
+                x = 164,
+                y = 6
+            },
+            size = {
```
Isn't the size given by the image? Do you need to specify it here?

Ideally you could do for the background the same thing you did for the foreground, i.e. make just a one pixel wide image and stretch it to the size the user specifies here. Then you'd need to put the left and right brackets into their own images, which would be a bit more work, but it would make it more flexible. Could consider that for a future update.

Morath86 replied:
I'll get around to that when I get into changing the colours, and allowing for changing sizes more easily.

z16 commented on addons/barfiller/barfiller.lua:
```diff
@@ -0,0 +1,88 @@
+-- Thanks to Byrth & SnickySnacks' BattleMod addon
+windower.register_event('addon command',function(...)
+    local commands = {...}
+    local first_cmd = table.remove(commands,1):lower()
```
Instead of doing this, there's a simpler way:

``` lua
windower.register_event('addon command', function(command, ...)
    command = (command or 'help'):lower()
    local args = {...}

    -- Rest here
end)
```

This will automatically make the first value the command and set it to `'help'` if it doesn't exist (your approach would currently result in a Lua error if the first command was missing).

Morath86 replied:
It was indeed, that has now been amended. Thanks.

z16 commented:
All of the current suggestions are just that, suggestions. The code itself looks fine and all the essentials are implemented. Let me know if you still want some time to implement those suggestions or if you want me to merge it as it is now.

Morath86 commented:
These are all great suggestions, and are helping me to better understand Lua. This is my first time working with it, and I appreciate all the criticism. I suspect I'll have another push ready in an hour or so.

Morath86 committed:
No longer specifying data\settings.xml when loading settings_table, really fixed PascalCase, defaulted command argument to help

Morath86 committed:
Added Job, Level, Current XP, Total XP, TNL, and Percent text labels

Morath86 committed:
Added Job, Level, Current XP, Total XP, TNL, and Percent text labels

Morath86 committed:
Now loads safely at character selection screen, and on character login

Morath86 committed:
Booleans added to limit chunk updates, and screen updates

Morath86 committed:
Removed old comment above 'prerender' event

Morath86 commented:
Added EXP/hr counter, fewer calculation updates, and now compatible at character selection screen. Ready for public usage at this point.

z16 commented on addons/barfiller/statics.lua:
```diff
@@ -0,0 +1,433 @@
+defaults = {
+    Bars = {
+        Background = {
+            Image = windower.windower_path..'addons\\barfiller\\data\\bar_bg.png',
```
I still advise very much against both specifying any paths in the settings file and having the icons in the `data` sub folder.

alphaONE2 replied:
I second this. I highly suggest this be changed before the addon is merged. Windower _will not_ overwrite or delete files in an addon's data directory. The directory doesn't belong to the addon, it belongs to the users. This was an intentional design decision so that addons could provide a safe location for user files that should not be overwritten during updates.

If you place images here you can't change them in the future. Things that an addon needs to function should not be in the data directory. Users should be able to delete the data directory and the addon should still be fully functional.

For the initial release I suggest using hard-coded paths (don't include them in settings) and moving the images into a separate directory, outside of the data directory. We can discuss the options you have for allowing customization for the next release with you on IRC if you want.

TL;DR: As this is you have _no_ upgrade path, you really cant change or update your images in the future if you leave it like this. For instance, you won't be able to make the bar change color dynamically, because the color is baked into the image. And, you wont be able to update that image, because it's in the data directory.

Morath86 replied:
I'll try and get on IRC tomorrow evening to discuss this one.

EDIT: I moved all 3 images to the root folder and changes the lines to
`Image = windower.addon_path..'file_names.png'`

alphaONE2 replied:
OK, that at least gives you an upgrade path that wont force users to reinstall the addon.

z16 commented on addons/barfiller/barfiller.lua:
```diff
@@ -0,0 +1,116 @@
+
+-- Make sure character is logged in, and loaded before initializing
+windower.register_event('load',function()
+    if windower.ffxi.get_info().logged_in and windower.ffxi.get_player() then
```
The second check is unnecessary, if `windower.ffxi.get_info().logged_in` is true then `windower.ffxi.get_player()` will also exist.

Morath86 replied:
Got it, thank you!

z16 commented on addons/barfiller/barfiller.lua:
```diff
@@ -0,0 +1,116 @@
+end)
+
+-- Delay initialize() for 10 seconds to allow game to download chunks
+windower.register_event('login',function()
```
This function should probably also call `initialize`. Currently `initialize` is only called when loading the addon, not when relogging.

Edit: Nevermind, I just saw that that's what the next function does. In which case there's a better way to write it:

``` lua
    initialize:schedule(10)
```

But why the 10 sec delay in the first place? Are you waiting for a specific packet? In which case you may wanna check for it with the `incoming chunk` event.

Morath86 replied:
I haven't figured out why, but without delaying the `initialize()` until you're fully logged in (on average 5-10 seconds on my slow 3Mbps connection) the player data would not populate. Exp would still calculate, but the job, and level details wouldn't load.

This is the current problem I'm looking into, trying to find out which chunk may be the relevant one to wait for.

z16 commented on addons/barfiller/barfiller.lua:
```diff
@@ -0,0 +1,116 @@
+
+-- If you're switching characters this will clear the addon from memory
+windower.register_event('logout',function()
+    windower.send_command('lua r barfiller')
```
Addons should be made to be able to handle relogging. There is not a lot that you need to consider for it. Reloading just unnecessarily spams the console and possibly removes previous unsaved settings.

Morath86 replied:
That was a (very poor) cheat on my part, I've changed it to hide all the `prim` and `text` objects in my next push.

z16 commented:
If you want I can merge this, but what I wrote about settings needs to be done now, or else it may screw up peoples' settings files in the future. Do you want to move that out of the settings? If so I'll wait with the merge, otherwise I'll do it.

Morath86 committed:
Moved images to addons root directory, cleaned up 'login' and 'logout' events

Morath86 committed:
Converted Images to Objects

Morath86 committed:
Settings adjusted for Image Objects, more getters and setters

Morath86 committed:
Initial commit, new Image library

Morath86 commented:
Added new Image library, updated prims to new use Objects, additional getters and setters added.

Morath86 committed:
Smooth animations added to 'prerender' event

Morath86 committed:
Moved :show() for Images and Texts to load_images() and create_text_box() respectively

Morath86 commented:
Smooth animation logic added to `prerender` event

Morath86 committed:
Fix err in prerender event when calc_exp_bar() returns nil

Morath86 committed:
Added Zone Change event

Morath86 commented:
Added err checking for when calc_exp_bar() returns nil

Morath86 committed:
Removed 10 second delay on login, forgetting to update windower.ffxi.get_player()

Morath86 committed:
Added frame_count back in to prevent flickering of text

Morath86 commented:
No more delay on startup/login, and re-added `frame_count` to prevent text flicker

alphaONE2 commented:
@Morath86, @z16: Are we still waiting on anything here or can this get merged?

Morath86 committed:
Added 'level up/down' events to make sure Player Data keeps updating

Morath86 commented:
@IryokuChevalier @z16 Unless you see any other critical issues, I believe it's ready. I haven't encountered any problems tonight, and just fixed my last known issue.

z16 commented on addons/libs/images.lua:
```diff
@@ -0,0 +1,344 @@
+return images
+
+--[[
+Copyright © 2013-2015, Windower
```
This should be `2015` only, since the library didn't exist before.

Morath86 replied:
Done.

z16 commented:
@IryokuChevalier @Morath86 I was really busy the last few days, and since this is adding a new library I believe it deserves some more scrutiny than just an addon commit. I'm going over it right now.

z16 commented on addons/barfiller/statics.lua:
```diff
@@ -0,0 +1,844 @@
+function set_fg_width(new_width)
+    fg_width = new_width
+    foreground_bar:size(new_width, get_fg_height())
+    if frame_count%15 == 0 then update_strings() end
```
I believe this should be done with a timer instead of a frame counter. If someone is playing with uncapped frame rate, this may still update seven times a second. If someone is lagging it may only update once a second or only every few seconds. If you did this with a timer (i.e. update if half a second has passed since the last time you did it) it would be more consistent:

``` lua
do
    local last_update = 0

    function set_fg_width(new_width)
        fg_width = new_width
        foreground_bar:size(new_width, get_fg_height())

        local now = os.clock()
        if now - last_update > 0.5 then
            update_strings()
            last_update = now
        end
    end
end
```

Edit:
I wrote this before the comment above. If you remove all of these lines I would put this directly into the `prerender` event. As a matter of fact, I would put it there regardless.

Morath86 replied:
Thanks, I'll jump on this tonight.

z16 commented on addons/barfiller/statics.lua:
```diff
@@ -0,0 +1,844 @@
+    local mh = windower.ffxi.get_info().mog_house
+    if mh then rested_bonus:show() else rested_bonus:hide() end
+end
+
```
Do you need any of the following functions? The benefit of the library was so you could get rid off all of these. The entire rest of the file could be removed, if I understand this correctly.

Morath86 replied:
As soon as I update the other functions that utilized them, I will remove the getters and setters.

z16 commented on addons/libs/images.lua:
```diff
@@ -0,0 +1,344 @@
+local meta = {}
+
+windower.prim.saved_images = {}
+local dragged
```
This variable isn't currently used. The texts library uses it to expose mouse events on the text boxes. I think it's worth having in this library as well.

Morath86 replied:
Okay I'll look into putting it back. I personally hated always missclicking the screen, and dragging the Text fields on the pointwatch addon, but it is a lot better easier than positioning with POS commands.

z16 replied:
You can disable dragging on a per-addon basis. There is a `draggable` attribute that can be passed to it as `true` or `false`, which defaults to `true`:

``` lua
text = texts.new('Some text', { draggable = false })
```

The TParty addon does this, for example, because it explicitly positions the text boxes. If you copied the API from the `texts.lua`, this functionality would be included in the `images.lua` library, so you could make _BarFiller_ use the custom X and Y coordinates and disable it, but other addons could still use it.

z16 commented on addons/libs/images.lua:
```diff
@@ -0,0 +1,344 @@
+
+    local settings = windower.get_windower_settings()
+    windower.prim.set_position(m.name, x, y)
+    m.settings.Pos.X = x
```
I'm almost certain this won't work, because Windower settings are not the same as addon settings, and they are all _snake_cased_.

Morath86 replied:
Woops, that would be a copy and paste fumble. They do work as it is, but I'll fix it for follow proper syntax.

Morath86 committed:
Fixed memory leak with calculations, Strings now update by factor of time instead of frames

Morath86 committed:
Fixed Text object settings, removed getters and setters, calc_exp_bar() renamed to calc_new_width(), create_text_box() renamed to load_text_box()

Morath86 committed:
Added dragging back in, Functions added images.width, images.height images.get_extents, images.draggable

Morath86 commented:
Fixed memory leak when calculating new width, renamed functions to be read easier, strings are now
updated by a factor of time instead of frames for a more consistent experience at any frame rate.

Morath86 committed:
Renamed images and text objects, added new toggle command Visible

Morath86 committed:
Rewrote settings, removed unused functions, added show() func for new Visible command
